pkg/kubelet/kubeclientmanager/kubeclient_manager.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2020 Authors of Arktos.
+Copyright 2020 Authors of Arktos - file modified.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubeclientmanager
+
+import (
+	"k8s.io/api/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type KubeClientManager struct {
+	tenant2api     map[string]int // tenant name -> api server id
+	tenant2apiLock sync.RWMutex
+}
+
+var ClientManager *KubeClientManager
+var kubeclientManagerOnce sync.Once
+
+func NewKubeClientManager() {
+	kubeclientManagerOnce.Do(
+		newKubeClientManagerFunc(),
+	)
+}
+
+func newKubeClientManagerFunc() func() {
+	return func() {
+		ClientManager = &KubeClientManager{
+			tenant2api: make(map[string]int),
+		}
+		klog.Infof("kubeclient manager initialized %v", ClientManager)
+	}
+}
+
+func (manager *KubeClientManager) RegisterTenantSourceServer(source string, ref *v1.Pod) {
+	if len(ref.Tenant) == 0 || !strings.HasPrefix(source, kubetypes.ApiserverSource) {
+		klog.Warningf("unable to register tenant source : tenant='%s', source='%s'", ref.Tenant, source)
+		return
+	}
+
+	manager.tenant2apiLock.Lock()
+	defer manager.tenant2apiLock.Unlock()
+
+	key := strings.ToLower(ref.Tenant)
+	if source == "api" {
+		klog.Warning("source is 'api', will only use kube client #0")
```

Comment:
Do we need a warning here? It is the backwards-compatible case. "Infof" might be enough.

Reply:
switched to Info